### ISSUE TITLE
Invalid settings from VK_EXT_layer_settings are not reported

### ISSUE THREAD

**Issue by elasota** [KhronosGroup/Vulkan-ValidationLayers#9028]
**Environment:**
 - OS: Windows 10
 - GPU and driver version: AMD Radeon RX 6700, driver 24.9.1
 - SDK or header version if building from repo: 1.3.296.0

**Describe the Issue**
If a setting is supplied VK_EXT_layer_settings that is not actually supported by the validation layer, no warning message is generated.  Since VK_EXT_layer_settings itself doesn't supply a reflection API to enumerate settings, there isn't really a way to validate that the settings being passed were correct.

Passing values with the wrong type also results in unusual behavior, e.g. passing "thread_safety" with the type set to VK_LAYER_SETTING_TYPE_INT32_EXT results in the setting being disabled.

**Expected behavior**
Passing an unsupported or incorrectly-typed setting should produce a warning.

**Valid Usage ID**
There does not appear to be a VUID for this.

**Comment by spencer-lunarg** [comment]
This a really good question and opened a lot of dialog

I created this issue https://github.com/KhronosGroup/Vulkan-Utility-Libraries/issues/254 but for the near term, I am going to just hand write it for the Validation Layers (and then swap it out for the VUL version) as I agree 100% you should be warned if you are setting invalid or  non-existent settings

**Comment by spencer-lunarg** [comment]
@elasota so added a fix to at least give you a warning if you provide a setting we don't have or the wrong type

As for a "reflection" it was brought up internally that during the design of `VK_EXT_layer_settings` that came up and the decision was made (which I agree with now) that if you don't know a setting you shouldn't just have code randomly try setting it

**Comment by elasota** [comment]
> As for a "reflection" it was brought up internally that during the design of `VK_EXT_layer_settings` that came up and the decision was made (which I agree with now) that if you don't know a setting you shouldn't just have code randomly try setting it

FWIW one of the cases I'm concerned about in wanting this is the setting not being supported by the SDK/driver version.  In that case, correct usage will still fail, and it would be nice to have a way (similar to layer/extension enumeration) to detect that the requested setting is unavailable and inform the user.

**Comment by spencer-lunarg** [comment]
So the issue becomes we currently don't have a mechanism to expose all the settings

We would need any layer that uses `VK_EXT_layer_settings` to provide a `vkGetLayerSettings` type query

>  I'm concerned about in wanting this is the setting not being supported by the SDK/driver version

I think this really becomes a task of the layer to not just pull away settings from under you. In Validation, we still support all the old setting and even some we regret the names, we will get a "deprecated" warning, but likely will never just "remove" them on you

**Comment by elasota** [comment]
> I think this really becomes a task of the layer to not just pull away settings from under you.

I mean in the situation where a new setting is added and isn't supported in older versions.

e.g. Suppose a setting is added (or expanded with new options) in the future and a program is written in a way that it wants to take advantage of that setting with a configuration option or command-line switch.  If another user/developer uses that program, the setting will not be available if the installed layer on their machine is too old, or if the program is being deployed to a device that doesn't support the setting in its implementation yet.

Assuming the program wants to configure the setting itself, it would be nice to detect that the setting is unavailable so the program can inform the user that there is a problem, or otherwise indicate that the setting is unavailable.

**Comment by elasota** [comment]
> I think this really becomes a task of the layer to not just pull away settings from under you.

I mean in the situation where a new setting is added and isn't supported in older versions.

e.g. Suppose a setting is added (or expanded with new options) in the future and a program is written in a way that it wants to take advantage of that setting with a configuration option or command-line switch.  If another user/developer uses that program, the setting will not be available if the installed layer on their machine is too old, or if the program is being deployed to a device that doesn't support the setting in its implementation yet.

Assuming the program wants to configure the setting itself, it would be nice to detect that the setting is unavailable so the program can inform the user that there is a problem, or otherwise indicate that the setting is unavailable.

**Comment by charles-lunarg** [comment]
VK_EXT_layer_settings was intentionally designed without reflection in mind, for the aforementioned reason of "if you don't already know what a setting is, you shouldn't be messing with it". 

The Vulkan Validation layer is a developer facing tool, which is expected to be enabled and configured by whoever was developing a Vulkan program directly. Use cases where the validation layer is used by a library or framework (as in, by non-Vulkan API developers or by end users) isn't the intended workflow, an was not a priority when designing VK_EXT_layer_settings as the Validation layer was a primary use case for the extension. 

Speaking as the Vulkan-Loader maintainer, I will add that a reflection API for VK_EXT_layer_settings is not a simple task. The reflection API would need to be available *before* you call `vkCreateInstance`. But in Vulkan, there isn't a way to 'enable' extensions before you create a VkInstance, creating a chicken-and-egg type problem. I don't think its *too* big of a problem to solve, it just creates more wrinkles than were desired, since the primary purpose of VK_EXT_layer_settings is to provide an in-API means of configuring layers.

**Comment by charles-lunarg** [comment]
Clarification: I don't mean to imply that invalid settings shouldn't be diagnosible by the application, but instead try to provide an explanation for why there isn't the "obvious" solution of having a reflection API. In fact, I wouldn't be opposed to that being added in the future, I just wanted it to be known that there was a high development cost to doing so. 

Better error reporting definitely should be in place, and discussion of that should ideally occur in the [Vulkan Utility Libraries issue](https://github.com/KhronosGroup/Vulkan-Utility-Libraries/issues/254)

**Comment by elasota** [comment]
Yeah I understand, I figured it may be viable since `vkEnumerateInstanceExtensionProperties` already performs a similar role of querying layer capabilities (although `VkExtensionProperties` doesn't have `sType`/`pNext` so there'd be no way to do it without a new function).

I realize it's not for end users, I was more interested in prospective contributors that may not be up-to-date but I guess grabbing an SDK package via the build system or something might be a better solution to that.

Anyway, I think I've made whatever case I have for it, so I don't really anything more to add.

**Comment by elasota** [comment]
Yeah I understand, I figured it may be viable since `vkEnumerateInstanceExtensionProperties` already performs a similar role of querying layer capabilities (although `VkExtensionProperties` doesn't have `sType`/`pNext` so there'd be no way to do it without a new function).

I realize it's not for end users, I was more interested in prospective contributors that may not be up-to-date but I guess grabbing an SDK package via the build system or something might be a better solution to that.

Anyway, I think I've made whatever case I have for it, so I don't really anything more to add.
